hax/test/integration/test_server.py

```diff
@@ -0,0 +1,157 @@
+import logging
+from base64 import b64encode
+from typing import List
+
+import pytest
+import simplejson
+from hax.log import setup_logging
+from hax.message import BroadcastHAStates
+from hax.motr import WorkPlanner
+from hax.motr.delivery import DeliveryHerald
+from hax.server import ServerRunner
+from hax.types import Fid, HAState, MessageId, ServiceHealth
+from hax.util import ConsulUtil
+
+
+@pytest.fixture(autouse=True)
+def logging_support():
+    setup_logging()
+    yield ''
+    logging.shutdown()
+
+
+@pytest.fixture
+def herald(mocker):
+    return DeliveryHerald()
+
+
+@pytest.fixture
+def planner(mocker) -> WorkPlanner:
+    def fake_add(cmd):
+        if hasattr(cmd, 'reply_to') and cmd.reply_to:
+            cmd.reply_to.put([MessageId(1, 42)])
+
+    planner = WorkPlanner()
+    mocker.patch.object(planner, 'add_command', side_effect=fake_add)
+    return planner
+
+
+@pytest.fixture
+def consul_util(mocker):
+    consul = ConsulUtil()
+    exc = RuntimeError('Not allowed')
+    mock = mocker.patch.object
+    mock(consul.kv, 'kv_get', side_effect=exc)
+    mock(consul.kv, 'kv_put', side_effect=exc)
+    mock(consul, 'get_local_nodename', side_effect=['localhost'])
+    mock(consul, 'get_hax_ip_address', side_effect=['192.168.0.28'])
+    return consul
+
+
+@pytest.fixture
+async def hax_client(mocker, aiohttp_client, herald, planner, consul_util,
+                     loop):
+    srv = ServerRunner(planner, herald, consul_util)
+    srv._configure()
+    return await aiohttp_client(srv.app)
+
+
+async def test_hello_works(hax_client):
+    resp = await hax_client.get('/')
+    assert resp.status == 200
+    text = await resp.text()
+    assert text == "I'm alive! Sincerely, HaX"
+
+
+@pytest.mark.parametrize('status,health', [('passing', ServiceHealth.OK),
+                                           ('warning', ServiceHealth.FAILED),
+                                           ('critical', ServiceHealth.FAILED)])
+async def test_service_health_broadcast(hax_client, planner, status: str,
+                                        health: ServiceHealth):
+    service_health = [{
+        'Node': {
+            'Node': 'localhost',
+            'Address': '10.1.10.12',
+        },
+        'Service': {
+            'ID': '12',
+            'Service': 'ios',
+            'Tags': [],
+            'Port': 8000,
+        },
+        'Checks': [
+            {
+                'Node': '12',
+                'CheckID': 'service:ios',
+                'Name': "Service 'ios' check",
+                'Status': status,
+                'Notes': '',
+                'Output': '',
+                'ServiceID': '12',
+                'ServiceName': 'ios',
+            },
+        ],
+    }]
+    resp = await hax_client.post('/', json=service_health)
+    assert resp.status == 200
+    assert planner.add_command.called
+    planner.add_command.assert_called_once_with(
+        BroadcastHAStates(
```

Comment:
How about saving the msg_ids (tags) too and waiting for them to be complete as we do it in some cases in hax?

Reply:
@mssawant That would be ideal. But there are several reasons why I didn't do that:
1. Waiting for a delivery will mean that there are multiple threads within the test (one is the test runner, another one - "Motr emulator"). Single thread implementation will just block forever, deliveries always happen from another thread.
2. aiohttp-aware tests use `async` and `await`s. It is a bit tricky to combine asynchronous code with threads. Especially given that helper threads will have to be started and stopped properly. In Python it is not that easy to force a thread to stop (e.g. if it hangs on a blocking syscall).

So I think that we'll need to earn some more expertise in asyncio to write the tests like you described without the fear to hang forever. So I decided to start from simpler scenarios.

Reply:
Okay, we can take it as an improvement later. But overall the changes look pretty awesome.